codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/OperationRuntimePluginGenerator.kt

```diff
@@ -106,4 +120,45 @@ class OperationRuntimePluginGenerator(
             },
         )
     }
+
+    private fun generateAuthOptions(
+        operationShape: OperationShape,
+        authOptions: List<AuthOption>,
+    ): Writable = writable {
+        if (authOptions.any { it is AuthOption.CustomResolver }) {
+            throw IllegalStateException("AuthOption.CustomResolver is unimplemented")
+        } else {
+            val authOptionsMap = authOptions.associate {
+                val option = it as AuthOption.StaticAuthOption
+                option.schemeShapeId to option
+            }
+            val authTrait: AuthTrait? = operationShape.getTrait() ?: codegenContext.serviceShape.getTrait()
+            withBlockTemplate(
+                "cfg.set_auth_option_resolver(#{DynAuthOptionResolver}::new(#{StaticAuthOptionResolver}::new(vec![",
+                "])));",
+                *codegenScope,
+            ) {
+                var atLeastOneScheme = false
+                for (schemeShapeId in authTrait?.valueSet ?: emptyList()) {
+                    val authOption = authOptionsMap[schemeShapeId]
+                        ?: throw IllegalStateException("no auth scheme implementation available for $schemeShapeId")
+                    authOption.constructor(this)
+                    atLeastOneScheme = true
+                }
+                if (operationShape.hasTrait<OptionalAuthTrait>()) {
+                    val authOption = authOptionsMap[noAuthSchemeShapeId]
+                        ?: throw IllegalStateException("missing 'no auth' implementation")
+                    authOption.constructor(this)
+                    atLeastOneScheme = true
+                }
+                if (!atLeastOneScheme) {
+                    throw IllegalStateException(
+                        "this client won't have any auth schemes " +
+                            "(not even optional/no-auth auth), which means the generated client " +
+                            "won't work at all for the ${operationShape.id} operation",
+                    )
```

Comment:
We should consider linking to smithy docs which explain auth schemes. I'm guessing this will come up for some people generating clients internally.

Reply:
Good call. Fixed in [040d3fe](https://github.com/awslabs/smithy-rs/pull/2808/commits/040d3fe6221420994aa93c776b32d37d87a3cb1d)